### DISCUSSION TITLE
How to add a new version of Python?

### DISCUSSION THREAD

**Discussion by davidreyg** [leokhoa/laragon#411] [Q&A]
Need help to replace / add python version

**Comment by Pen-y-Fan** [comment]
Python 3.10 is included with Laragon 6.

- [How to update? #405](https://github.com/leokhoa/laragon/discussions/405)
- [How to set up Laragon on a new Windows computer (part 1) - Install Laragon](https://pen-y-fan.github.io/2022/11/07/how-to-set-up-laragon-on-a-new-windows-computer/)

With previous versions, it had to be added using Laragon > Tools > Quick add  > python-3.6

## Manual Install

To manually install a new version see my blog [How to update the PHP version in Laragon](https://pen-y-fan.github.io/2023/01/15/how-to-update-the-php-version-in-laragon/), as the process is similar.

- Download the Windows embeddable package from <https://www.python.org/downloads/windows/>
- Unzip the package
- Copy the folder to laragon\bin\python

## Automated Install

It's also possible to automate the process by adding the following to laragon\usr\packages.conf (Laragon menu > Tools > Quick Add > configuration)

```text
# Python
python-3.11=https://www.python.org/ftp/python/3.11.2/python-3.11.2-embed-amd64.zip
```

Python 3.11 can then be automatically downloaded and installed using the Laragon menu > Tools > Quick Add > python-3.11.

## Switch version

Laragon Menu > Python > Version > python-3.11-embedded-amd64

(If you do not like the switch version name rename the folder c:\laragon\bin\python\\**python-3.11-embedded-amd64** to **python-3.11**)

![Python 3.11](https://user-images.githubusercontent.com/40126936/219766824-ec951aed-6927-4570-8e2d-9bde11b5c13a.png)

Open the **Terminal** and the selected version (Python 3.11) is available.

![Cmder](https://user-images.githubusercontent.com/40126936/219767275-fa80fdd4-6fe2-4810-82a1-a9a35097d392.png)

**Reply by aldinal21** [reply]
if i have laragon python, why does the pip is not found, and the venv is no module

**Reply by Pen-y-Fan** [reply]
Please try this and let us know how you get on:

- [How to Install PIP on Windows ?](https://www.geeksforgeeks.org/how-to-install-pip-on-windows/)

**Reply by Akbarsait** [reply]
@Pen-y-Fan Thanks for the instructions. I followed the automated approach and installed Python 3.12 
# Python
python-3.12=https://www.python.org/ftp/python/3.12.3/python-3.12.3-amd64.exe

However, after installing and switching the version to 3.12, the version is still showing as 3.10 

`PS C:\laragon> python --version
Python 3.10.6`

Is there anything I'm missing?

**Reply by Akbarsait** [reply]
Team, - Anyone assist with the above request please @Pen-y-Fan @davidreyg @reinierbutot

**Reply by Pen-y-Fan** [reply]
I ended up using the full installation of Python, which included PIP etc. I installed each version in the laragon folder path:

- C:\laragon\bin\python\python-3.11 
- C:\laragon\bin\python\python-3.12

Laragon will then allow the versions to be switched.

<img width="503" alt="image" src="https://github.com/leokhoa/laragon/assets/40126936/01336b81-aea3-446e-9c36-ef159d63ab37">

![image](https://github.com/leokhoa/laragon/assets/40126936/e8a3262b-2dfe-478b-93f3-05e3d2378f22)

**Reply by Pen-y-Fan** [reply]
When installing Python select custom install and choose **Customize install location** by clicking **Browse** and navigate to your laragon directory > bin > python, make folder **python-3.12** and select it:

![image](https://github.com/leokhoa/laragon/assets/40126936/195bed81-65ec-4cbc-8f42-5716cb1b7bce)

You can now switch versions and Laragon will automatically update the path:

![image](https://github.com/leokhoa/laragon/assets/40126936/65994f1f-8e5c-4741-a56c-22181d4ceeb9)

If you have any problems check the python.exe is in the correct directory:

```sh
C:\laragon\bin\python\python-3.12\python.exe --version
Python 3.12.3
```

If python.exe isn't in the correct directory, use the Windows add or remove programs to uninstall it and use the installer to install it again.

Edit: Changed app to add or remove programs

**Reply by Akbarsait** [reply]
Perfect. Custom install solves the problem.

**Comment by mrhili** [comment]
what about the pip
because no ensure pip no pip even i downloaded pip from web no modue named pip

**Comment by luisAntonioLAGS** [comment]
1. Enter the following link
https://winpython.github.io/
2. Select the version and click on the repository where you want to download it, either github or sourceforge
![imagen](https://github.com/user-attachments/assets/de7dd274-9000-486e-ad61-d49d74dd68b3)
3. Download the versions that have the word dot at the end, since these are the versions that do not contain any libraries
![imagen](https://github.com/user-attachments/assets/d2a15db3-c8c8-4233-8c92-f8f95bd7453c)
4. Unzip it in the directory of your choice, in this case D:\Downloads\python12
![imagen](https://github.com/user-attachments/assets/ac2791d5-73b1-4458-98ef-ba0f8a78e5e7)
5. Open the folder you selected to unzip it
![imagen](https://github.com/user-attachments/assets/02953a9b-85d4-4e8f-aa73-8c46f609f791)
6. Enter the WPy64-31241 folder
![imagen](https://github.com/user-attachments/assets/a25299cd-c893-4837-80f7-e48b85182b66)
7. Move or copy the python-3.12.4.amd64 folder to the folder where the python versions are located in laragon
![imagen](https://github.com/user-attachments/assets/d580c70e-0363-4254-a4c5-1d2f0b585d81)
9. Once the folder has been moved, you can select the new version of python in the laragon interface
![imagen](https://github.com/user-attachments/assets/05de616f-3b9b-47bf-9cc4-7b138dccc17d)
![imagen](https://github.com/user-attachments/assets/e8ad7792-3cc7-400c-8ccd-086e924dea6f)
